TouchPortalAPI/TpToPy.py

```diff
@@ -0,0 +1,140 @@
+import json
+from pathlib import Path
+
+class TpToPy():
+    def __init__(self, entry):
+        self.entry = json.loads(Path(entry).resolve().read_text())
+    
+    def getPluginId(self):
+        return self.entry.get("id", "")
+
+    def __convertData(self, data):
+        newData = {}
+        if isinstance(data, list):
+            for item in range(len(data)):
+                newData[item+1] = data[item]
+        return newData
+
+    def __convertFormat(self, actionFormat, data):
+        newFormat = actionFormat
+        for item in range(len(data)):
+            if data[item]['id'] in actionFormat:
+                newFormat = newFormat.replace("{$" + data[item]['id'] + "$}", "$[" + str(item+1) + "]")
+
+        return newFormat
+
+    def generateInfo(self):
+        generatedInfo = {}
+        infoKeys = ["sdk", "version", "name", "id", "configuration", "plugin_start_cmd_windows", "plugin_start_cmd_linux", "plugin_start_cmd_mac", "plugin_start_cmd"]
+
+        for key in self.entry.keys():
+            if key in infoKeys:
+                generatedInfo[key] = self.entry[key]
+
+        return generatedInfo
+    
+    def generateSettings(self):
+        generatedSetting = {}
+        if not self.entry.get("settings", True): return generatedSetting
+
+        settings = self.entry.get("settings", [])
+
+        if isinstance(settings, list):
+            for setting in range(len(settings)):
+                generatedSetting[setting+1] = settings[setting]
+        return generatedSetting
+
+    def generateStates(self):
+        generatedState = {}
+        categories = self.entry["categories"]
+        stateNum = 1
+        for category in range(len(categories)):
```

Comment:
I don't really understand why you loop like this.... it's inefficient and harder to understand. Simply:
```py
for category in categories:
    if states := category.get('states') and isinstance(states, list):
        for state in states:
       .....
```

Reply:
Also overall wouldn't it be more efficient to loop over each category once and then from there call the various methods which generate stuff from that category... like states/actions/connectors?

Reply:
True, Never thought about that...

Reply:
Thanks for the suggestion